add string/number operator subtypes to dashboard parameter filters

daltojohnso opened the pull request:
This PR adds new types of parameter widgets that are similar to the various operators found in Question filters. See related issue https://github.com/metabase/metabase/issues/15055 for a list of the added parameter types.

<img width="342" alt="Screen Shot 2021-03-23 at 4 48 32 PM" src="https://user-images.githubusercontent.com/13057258/112233431-aec10380-8bf7-11eb-99a5-c9d02da95f81.png">
<img width="338" alt="Screen Shot 2021-03-23 at 4 48 56 PM" src="https://user-images.githubusercontent.com/13057258/112233436-b08ac700-8bf7-11eb-89e1-770b1a3452a4.png">

Parameters can be added to Dashboards as well as Native Questions. Besides there being new parameter options to choose from the general behavior of parameters in both places should be the same.

<img width="351" alt="Screen Shot 2021-03-23 at 4 47 27 PM" src="https://user-images.githubusercontent.com/13057258/112233480-bf717980-8bf7-11eb-86b3-217a66af417f.png">

**What's left out of this PR**
- The UI/UX of popovers in the Native Question sidebar should be replaces with inline components.
- The "all-options" parameter widget options will be added in a subsequent PR

daltojohnso committed:
Remove location sub-categories

These sub-categories are only for filtering the list of options
when mapping a parameter filter to a field. Since we are introducing
operator types as a sub-category of location, city/zip/etc. just get in
the way.

daltojohnso committed:
add number section + number/string operator subtypes

Light refactor of meta/Dashboard changes

rmv 'all-options' options (for now)

daltojohnso committed:
add/update parameter type icons

daltojohnso committed:
pass operator to ParameterFieldWidget + show input per operator field

daltojohnso committed:
Add operator helper fns that aren't dependent on fields/tables

daltojohnso committed:
Make operator prop optional

fix date filter err

daltojohnso committed:
add combined  name for native question filter widget type list

Otherwise, a field that matches both "Location" and "Category"
options will show duplicate "Starts with" options, etc. Now,
that'll look like "Category - Starts with" and "Location - Starts with"

daltojohnso committed:
correct some unused prop/arg passing

daltojohnso committed:
Convert location/category parameter types to string for query

location/category don't mean anything to BE but we use them for
"reasons" on the FE. Reasons are legacy reliance on unique-ness of
the parameter.type value, primarily.

daltojohnso committed:
add max-width to PopoverPicker

daltojohnso committed:
rmv unused value

daltojohnso committed:
use combinedName on dashboard parameters

daltojohnso committed:
Update Cypress tests to reflect new parameter flow

fix cypress dashboard parameter tests

Fix more cy tests

dpsutton committed:
operators in backend

dpsutton committed:
Remove errant tap>

dpsutton committed:
Docstrings and differing numbers in tests in some dbs

dpsutton committed:
Make unary private so docstring checker ~shuts-up~ is satisfied

dpsutton committed:
Don't parse arguments to operators params

they were coming in just fine from the FE as numeric or string
types. no need to ensure strings everywhere and parse here

daltojohnso committed:
fix parameter to mbql code

daltojohnso committed:
Ensure = operator filter popovers have no label

This is to match "old" style of parameter popover

dpsutton committed:
Namespace doc and remove unnecessary comment

daltojohnso committed:
Don't call fk/joinAlias on ExpressionDimension

The methods don't exist on ExpressionDimension class.

This doesn't make them work (yet), but it prevents the app
from crashing.

dpsutton committed:
tap>-spy in dev

dpsutton committed:
first pass at substitution of new operators in native

dpsutton committed:
Docstring on wrap-value-literals-in-mbql to appease the gods

dpsutton committed:
variadic equality operators (string/= number/=)

daltojohnso committed:
move functions out of component file

daltojohnso committed:
Pass parameter object to tag editor for use in default input

We should inline this input eventually because it looks ugly.

daltojohnso committed:
map parameters in Questions to correct type

daltojohnso committed:
continue to pass janky fake parameter for text/number tags

dpsutton committed:
mongo native substitution

dpsutton committed:
variadic not-equals for string and number

dpsutton committed:
Docstring and use correct function to make errors

daltojohnso committed:
add number/between dash param cy test

daltojohnso committed:
Update function name to better reflect behavior

daltojohnso committed:
Add unit tests for paramer/operator util fns

add unit tests for parameter util functions

add unit test for operator util fns

mazameli commented:
Found a funny bug where the X is floating here in the middle of the input when the label is long and the filter is active:

![image](https://user-images.githubusercontent.com/2223916/112067561-7fda5d00-8b25-11eb-9887-4b3f31db93f8.png)

mazameli commented:
@daltojohnso, I pushed just now a small tweak to the placeholder text for new filter widgets. One thing I think we should do that I couldn't figure out how to do quickly is:

For Location and Category widgets, I think the placeholder label for the "matches exactly" widget should just be `Location` and `Category` respectively. 

(I may also tweak the name of that widget in the new filter selection popover to something like "Dropdown," further emphasizing that it'd be nice to not have to have the `option.name` concatenated onto the end for that particular widget.)

daltojohnso committed:
add variadic string 'is not' param operator option

daltojohnso committed:
Modify operator parameter display labels

don't append 'matches exactly' to location/category parameters

label tweaks

Update cy tests to reference correct label name

rmv it.only

daltojohnso commented:
@mazameli -- opened a separate fix for the parameter widget width weirdness: https://github.com/metabase/metabase/pull/15292

I also made the change to remove "matches exactly" from category/location labels.

dpsutton committed:
Desugar mongo parameters mbql

desugaring makes for a bit more verbose query but that's ok.This
change was done to ensure that we negated regexes in a correct way,
and to do so we always return the string version. This ensures that it
can be json/generate-string'd for native parameters or left as
datastructures and sent to monger

daltojohnso commented on frontend/src/metabase/meta/Parameter.js:
```diff
@@ -356,19 +455,71 @@ export function parameterToMBQLFilter(
+  }
+}
+
+export function mapUITypeToParameterType(parameter) {
```
There's a decent amount of code on the FE that expects `parameter.type` to be unique, so this is my workaround, to replace `location` and `category` with `string` to map to what the backend expects the type to be. It feels a little gross, but here are the benefits of this approach:

1. It's an easy way to handle existing `parameters` with types like `category` and `location/city`. They are both converted to `string/=`.
2. We will need some sort of function like this in the near future to make the `all-options` widget work.

daltojohnso commented on frontend/src/metabase/lib/schema_metadata.js:
```diff
@@ -470,6 +470,28 @@ const MORE_VERBOSE_NAMES = {
+  return !!FIELD_FILTER_OPERATORS[operatorName];
+}
+
+export function getOperatorByTypeAndName(type, name) {
```
"operators" aren't really first-class objects in the sense that there seem to be multiple different representations of what they are within this `schema_metadata` file. The pre-existing `getFilterOperators` is _almost_ useful, but depends on `field` and `table` args, so I've added this new getter function to bag up what operator metadata exists in this file without a dependency on anything besides a given `type` and `name`.

daltojohnso commented on frontend/src/metabase/parameters/components/widgets/ParameterFieldWidget.jsx:
```diff
@@ -138,42 +157,63 @@ export default class ParameterFieldWidget extends Component<*, Props, State> {
+              <div className="text-bold mb1">{verboseName}...</div>
+            )}
+
+            {_.times(numFields, index => {
```
This code resembles `DefaultPicker` quite a bit. Eventually, I hope to reconcile differences between the two implementations but that'll take some additional refactoring of those components, as they expect a card/question instance, which we don't have here.

daltojohnso commented on frontend/src/metabase/query_builder/components/template_tags/TagEditorParam.jsx:
```diff
@@ -241,11 +241,20 @@ export default class TagEditorParam extends Component {
+                        tag["widget-type"] ||
+                        (tag.type === "date" ? "date/single" : null),
+                    }
+              }
```
is a `tag.type` of `"dimension"` sufficient to mean we are intending to filter on a specific field or are there other things I need to concern myself with here?

Not really clear on the whole `tag` thing but as it stands in order to make the new parameter types work we need to pass in the real `parameter` object vs this fake `{type: ...}` object we were passing in before -- but we still need the `{type}` object to render `TextWidget` in scenarios where we're creating a text filter.

camsaul replied:
pretty sure `dimension` always means a Field

paulrosenzweig replied:
Tags are anything inside `{{...}}` in a native query. They can be "normal" variables (text, number, date, or field filters), snippets, or cards. Field filters' type is "dimension" and they always refer to a single field.

daltojohnso commented:
Some recordings of basic functionality

https://user-images.githubusercontent.com/13057258/112234122-09a72a80-8bf9-11eb-9aae-523e29bbff7a.mov


https://user-images.githubusercontent.com/13057258/112234132-1035a200-8bf9-11eb-8a49-7db7e5fef907.mov

dpsutton commented on modules/drivers/mongo/test/metabase/driver/mongo/parameters_test.clj:
```diff
@@ -235,4 +300,84 @@
+                      #{"bob" "tupac"}
+                      ;; most of these are nil as most records don't have a username. not equal is a bit ambiguous in
+                      ;; mongo. maybe they might want present but not equal semantics
+                      (into #{} (map second)
```
This comment is worth highlighting. `[:does-not-contain 'field "bob"]` returns true for all records with the field which don't match "bob" as well as records without 'field. This might be a bug in our mbql where we should emit that the field exists and it does not contain "bob"?

camsaul replied:
It's a little tricky because MongoDB doesn't have a formal schema the same way SQL databases do. I'm not really sure what we should do with filter operators like these for DBs with dynamic schemas if the key doesn't exist at all. I think we should open an issue and think about it there... I wouldn't change the existing behavior for this PR

dpsutton commented on src/metabase/driver/common/parameters/operators.clj:
```diff
@@ -0,0 +1,79 @@
+          [(unary param-type) field' a]
+
+          (contains? variadic param-type)
+          (into [(variadic param-type) field'] param-value)
```
since we verify type and arity tthis whole thing could essentially be this one line `(into [(translate param-type) field'] param-value)`. I left it as such in case we do need to do more transforms. Originally I was going to have a var of `negated` values and return `[:not (to-clause (update param :type negate))]` but mbql already had sugar for that (`:does-not-contain` -> `[:not [:contains ...]]`).

dpsutton commented on src/metabase/driver/sql/parameters/substitution.clj:
```diff
@@ -243,8 +233,39 @@
+                 (wrap-value-literals/wrap-value-literals-in-mbql)
+                 (sql.qp/->honeysql driver/*driver*)
+                 (hsql/format-predicate))]
+        {:replacement-snippet snippet, :prepared-statement-args (vec args)})
```
This i think should be the way the rest of these work. MBQL is quite nice to target (as evidenced by the essentially single line that accomplishes all of the `:string/` and `:number/` operators. Targeting this and then compiling mbql to the target dialect seems far preferable than hand-rolling another sql compiler here.

Certainly with emitting mongo i benefited from the knowledge in that rather than creating my own. And perhaps the hand optimized code might be a bit faster, but that's unproven, and a good motivator to optimize mbql in general to get this speedup everywhere.

camsaul replied:
Agreed. We should refactor the old stuff to use the new pattern (either as part of this PR or as a follow-on)

dpsutton commented on src/metabase/query_processor/middleware/parameters/mbql.clj:
```diff
@@ -28,17 +38,14 @@
-    ;; otherwise convert to a Long
     :else
-    (Long/parseLong param-value)))
+    (to-numeric param-value)))
```
I ended up not parsing anything here. I suspect we need to figure out how to do field-filters from query strings and ensure it works

dpsutton committed:
Cleanup stale comments and fixup docstring for consistency

dpsutton commented:
We need to verify that these field-filters work when used in the query string. I'm not sure how to exercise this actually.

camsaul commented on frontend/src/metabase/meta/Parameter.js:
```diff
@@ -21,15 +21,109 @@ import type Field from "metabase-lib/lib/metadata/Field";
+    //   description: t`Contains all of the above`,
+    // },
+  ],
+  string: [
+    {
+      operator: "=",
+      name: t`Dropdown`,
+      description: t`Select one or more values from a list or search box.`,
+    },
+    {
+      operator: "!=",
+      name: t`Is not`,
+      description: t`Exclude one or more specific values.`,
+    },
+    {
+      operator: "contains",
+      name: t`Contains`,
+      description: t`Match values that contain the entered text.`,
+    },
+    {
+      operator: "does-not-contain",
+      name: t`Does not contain`,
+      description: t`Filter out values that contain the entered text.`,
+    },
+    {
+      operator: "starts-with",
+      name: t`Starts with`,
+      description: t`Match values that begin with the entered text.`,
+    },
+    {
+      operator: "ends-with",
+      name: t`Ends with`,
+      description: t`Match values that end with the entered text.`,
+    },
+    // {
+    //   operator: "all-options",
+    //   name: t`All options`,
+    //   description: t`Users can pick from any of the above`,
+    // },
+  ],
```
I think this sort of stuff should go in the shared backend/frontend Clojure/ClojureScript lib in the future... we don't really need to list all the different filter types in both the FE and backend if we can just do it in one place

daltojohnso replied:
Definitely agree

camsaul commented on frontend/src/metabase/meta/Parameter.js:
```diff
@@ -95,33 +185,36 @@ function fieldFilterForParameter(parameter: Parameter) {
   parameterType: ParameterType,
 ): FieldPredicate {
-  const [type] = parameterType.split("/");
+  const [type] = splitType(parameterType);
   switch (type) {
     case "date":
       return (field: Field) => field.isDate();
     case "id":
       return (field: Field) => field.isID();
     case "category":
       return (field: Field) => field.isCategory();
+    case "location":
+      return (field: Field) =>
+        field.isCity() ||
+        field.isState() ||
+        field.isZipCode() ||
+        field.isCountry();
+    case "number":
+      return (field: Field) => field.isNumber();
```
this especially would be a good candidate for the shared lib. "What are the filter types I can use on a Field?" is something we should define there. It's useful in the backend too for stuff like generating automagic dashboards

camsaul replied:
not saying we should change all this stuff that was already written -- especially if it's working and has tests -- but in the future we should put this stuff in the shared lib

camsaul commented on frontend/test/metabase/meta/Parameter.unit.spec.js:
```diff
@@ -95,18 +100,169 @@ describe("metabase/meta/Parameter", () => {
+    });
+
+    it("should default the operator to `=`", () => {
+      expect(
+        stringParameterValueToMBQL(
+          { type: "category", value: ["1", "2"] },
+          null,
+        ),
+      ).toEqual(["=", null, "1", "2"]);
```
This 100% should go in the shared query lib, to the point that I'd say it's even worth moving it over as part of this PR. (The other stuff I mentioned earlier would have been nice, but not so nice that it's worth redoing). Long-term, our goal is for the frontend JS code not deal with MBQL directly at all, but rather have all the interaction with it happen thru the shared lib. Porting all of `metabase-lib` (and other parts of the FE client that deal with MBQL, like this one, that aren't part of `metabase-lib`) to the shared lib is going to take some time, but we can at least make sure new code goes to the shared lib going forward so we don't make the project harder

camsaul replied:
This code would also be 10x easier to write and test in Clojure

daltojohnso replied:
If it makes you feel any better, these MBQL-related functions already existed--I added a bit of additional logic to them and also added unit tests since there were none before.

`stringParameterValueToMBQL` and `numberParameterValueToMBQL` now depend on code in `schema_metadata.js` so we'd need to port a bit of that as well. It'd add a decent amount of bulk to this PR.

I'm not against porting but I'd definitely need some help as the cljs side of things is still all new to me.

camsaul commented on modules/drivers/mongo/src/metabase/driver/mongo/parameters.clj:
```diff
@@ -108,6 +116,22 @@
+                                    ;; desugar only impacts :does-not-contain -> [:not [:contains ... but it prevents
+                                    ;; an optimization of [:= 'field 1 2 3] -> [:in 'field [1 2 3]] since that
+                                    ;; desugars to [:or [:= 'field 1] ...].
+                                    (mbql.u/desugar-filter-clause)
```
You don't need to wrap these... you can just do `mbql.u/desugar-filter-clause`

dpsutton replied:
heh. every time i write these i vacillate back and forth between wrapped and unwrapped. i'll stick with unwrapped from now on

camsaul commented on modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj:
```diff
@@ -289,7 +289,10 @@
-(defmulti ^:private compile-filter mbql.u/dispatch-by-clause-name-or-class)
+(defmulti compile-filter
+  "Compile an mbql filter clause to datastructures suitable to query mongo. Note this is not the whole query but just
+  compiling the \"where\" clause equivalent."
```
Can you add `:arglists` metadata to the `defmulti`? It's nice to be able to know what arguments the multimethod takes without having to go hunt down a `defmethod` form for it. Editors can show this info too

camsaul commented:
Backend stuff LGTM. I think we should port some of the FE stuff not directly related to UI to the shared lib tho (especially the stuff that compiles filters to MBQL). It will be 100x easier to write and test in Clojure, and we can make use of it for stuff like Automagic Dashboards and the like going forward. 

We want to have all MBQL-related code in the shared lib eventually, and step one to realizing that dream is to stop writing new JS code that deals with MBQL

dpsutton committed:
Arglists metadata on defmulti and denude some threaded forms

paulrosenzweig commented on frontend/src/metabase/meta/Parameter.js:
```diff
@@ -145,7 +238,7 @@ export function variableFilterForParameter(
 
 function tagFilterForParameter(parameter: Parameter): TemplateTagFilter {
-  const [type, subtype] = parameter.type.split("/");
+  const [type, subtype] = splitType(parameter);
```
I think this should include number params, but we probably also want to use subtype. A betweens filter can't be connected to a number template tag but single arity subtypes can.

daltojohnso replied:
I ended up preventing usage of any operators besides `=` when mapping to tags. Do we need it now?

paulrosenzweig replied:
That should work in most cases. Longer term, we probably want to allow other subtypes since the variable/tag might be used that way in the query. e.g. I might want to map a "starts with" parameter to a query with `column LIKE {{tag}} || '%'`. We can't know that's how `tag` is used, so the right behavior is probably to be too permissive and allow it if it could make sense.

daltojohnso replied:
Agree that it'd be nice to allow this going forward but it requires some additional UI work on the native q side of things, so I'll keep it out of this PR for now.

paulrosenzweig reviewed (approved):
Changes look great! 👏 👏 👏 

A few QA things I saw:

1. I think we need to refine the warning when mapping to a native parameter that isn't a field filter. Currently, it always says: "This field only accepts a single value because it's used in a SQL query." That doesn't make sense for some subtypes (e.g. "contains"). 
2. The QB filter options need to include all the options that a parameter can now be converted to. It looks like ZIP codes don't support "starts with" in the QB. The MBQL is converted correctly, but the UI is broken:
![image](https://user-images.githubusercontent.com/691495/112348297-129e0780-8c9e-11eb-86c9-a9e35ebab2f4.png)
3. Mapping a "starts with" filter to a template tag breaks the parameter generation for connected GUI questions. The value is no longer wrapped in an array.

daltojohnso committed:
add single arity number tag predicate to variable filter

daltojohnso committed:
add Location operators to fix parameter<->filter mapping

For question filters to work we need for the new parameter operators
to be supported by "location" fields in all areas of the app.

daltojohnso committed:
Don't show coords for param number widgets

I don't think we want to support all the various number operators when
dealing with coordinates, so in order to avoid that I'm preventing the
mapping of number parameter operator to coord fields.

daltojohnso committed:
prevent mapping of tags to non-equal operators

while possibly useful to end users, this needs more UI work on the
native question side of things.

daltojohnso commented:
@camsaul I definitely will need some help with porting things to CLJS. That definitely won't be easier for me to write 😛 though maybe one day in the future. 

The functions I modified and wrote tests for (`stringParameterValueToMBQL` and `numberParameterValueToMBQL`) depend on logic in `schema_metadata.js` now so I imagine that needs to get ported as well.

Also, IDK how to write tests for cljs code. Didn't see any set up in the `shared` folder.

daltojohnso committed:
Ensure parameter values are wrapped in an array

When an = operator is mapped to a field AND a tag, it ends up not being
wrapped in an array due to the TextWidget (I think).

ensure parameter value is an array

ensure number params have an array value

dpsutton committed:
Sort imports correctly

clojure-lsp used to do this incorrectly (sorting `[` before `j`) and
that has now been fixed
